homeassistant/components/ted5000/sensor.py

```diff
@@ -49,85 +165,148 @@ def setup_platform(
+            current_tou = 0
+            current_tou_str = "Not Configured"
+        else:
+            current_tou = int(doc["LiveData"]["Utility"]["CurrentTOU"]) + 1
+            current_tou_str = doc["LiveData"]["Utility"]["CurrentTOUDescription"]
+
+        self.data_utility = {
+            ENTITY_MTUS: mtus,
+            ENTITY_RATE: current_rate / 100000,
+            ENTITY_DAYSLEFT: days_left,
+            ENTITY_PLANTYPE: plan_type_str[plan_type],
+            ENTITY_TIER: current_tier,
+            ENTITY_TOU: current_tou,
+            ENTITY_TOUDESC: current_tou_str,
+            ENTITY_CARBONRATE: carbon_rate / 100,
+            ENTITY_METERREAD: read_date,
+        }
 
+        return mtus
 
-class Ted5000Gateway:
-    """The class for handling the data retrieval."""
+    return 0
+
+
+class Ted5000Data:
+    """Collect data from the ted5000 gateway."""
 
     def __init__(self, url):
         """Initialize the data object."""
         self.url = url
-        self.data = {}
+        self.mtus = 0
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
-        """Get the latest data from the Ted5000 XML API."""
+        """Get the latest data from ted5000."""
+        # Update data
+        self.mtus = get_ted5000(self)
 
-        try:
-            request = requests.get(self.url, timeout=10)
-        except requests.exceptions.RequestException as err:
-            _LOGGER.error("No connection to endpoint: %s", err)
-        else:
-            doc = xmltodict.parse(request.text)
-            mtus = int(doc["LiveData"]["System"]["NumberMTU"])
 
-            for mtu in range(1, mtus + 1):
-                power = int(doc["LiveData"]["Power"]["MTU%d" % mtu]["PowerNow"])
-                voltage = int(doc["LiveData"]["Voltage"]["MTU%d" % mtu]["VoltageNow"])
+class Ted5000SensorEntity(SensorEntity):
+    """Representation of the current ted5000 data."""
 
-                self.data[mtu] = {
-                    POWER_WATT: power,
-                    ELECTRIC_POTENTIAL_VOLT: voltage / 10,
-                }
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        gateway,
+        mtu,
+        description: SensorEntityDescription,
+        name,
+    ):
+        """Initialize the sensor."""
+        self.entity_description = description
+        if mtu > 0:
+            self._attr_name = f"{name} MTU{mtu} {description.name}"
+        else:
+            self._attr_name = f"{name} Utility {description.name}"
+        self.gateway = gateway
+        self._mtu = mtu
+
+    def update(self):
+        """Get the latest data from Ted5000 and update the state."""
+        self.gateway.update()
+        if self._mtu > 0:
+            if self.entity_description.key in self.gateway.data[self._mtu]:
+                self._attr_native_value = self.gateway.data[self._mtu][
+                    self.entity_description.key
+                ]
+        if self.entity_description.key in self.gateway.data_utility:
```

Comment:
I think you can use `elif` here because there's not a scenario when you want to do both.

Reply:
You're right, I'll make the change...